test/controllers/sessions_controller_test.rb

```diff
@@ -1,7 +1,86 @@
 require 'test_helper'
 
 class SessionsControllerTest < ActionController::TestCase
-  # test "the truth" do
-  #   assert true
-  # end
+
+  class Create < SessionsControllerTest
+    def set_auth_mock(provider)
+      request.env['omniauth.auth'] = OmniAuth.config.mock_auth[provider].dup
+    end
+
+    setup do
+      set_auth_mock :github
+    end
+
+    test 'with no OAuth data redirects to root' do
+      request.env['omniauth.auth'] = nil
+
+      get :create, provider: :github
+
+      assert_response :redirect
+      assert_redirected_to root_path
+      refute_nil flash[:notice]
+    end
+
+    test 'OAuth login redirects to pull requests page' do
+      get :create, provider: :github
+
+      assert_response :redirect
+      assert_redirected_to pull_requests_path
+      assert_nil flash[:notice]
+    end
+
+    test 'first-time OAuth login creates a new user' do
+      assert_difference(lambda{ User.count }, 1) do
+        get :create, provider: :github
+      end
+
+      # Additional logins do not create new users
+      assert_no_difference(lambda{ User.count }) do
+        get :create, provider: :github
+      end
+    end
+
+    test 'uninvited users have unknown role' do
+      set_auth_mock :uninvited
+
+      get :create, provider: :github
+
+      assert_equal 'unknown', User.last.role
```

Comment:
Yeah, this is probably one of the things I shouldn't have done when building the authorization system. Having the 'unknown' role seemed to make sense when I was doing it, but we probably could have just made it impossible to have a user that didn't have a specific role.

However, we've already got this setup in place and it doesn't seem like it will break anything to keep it. Yes, I do plan to restrict all functionality for users with the unknown role.